ltm/inserted_context_agent.py

```diff
@@ -0,0 +1,577 @@
+{queries}  
+"""
+
+        prompt = """Here are a number of interactions, each is given a number:
+{passages}         
+*********
+
+Each of these interactions might be related to the general situation below. Your task is to judge if these interaction have any relation to the general situation.
+Filter out interactions that very clearly do not have any relation. But keep in interactions that have any kind of relationship to the situation such as in: topic, characters, locations, setting, etc.
+
+SITUATION:
+{situation}
+
+Express your answer in this JSON: 
+[
+    {{
+        "number": int  // The number of the interaction.
+        "justification": string  // Why the interaction is or is not related to the situation.
+        "related": bool // Whether the interaction is related to the situation.
+    }},
+    ...
+]
+"""
+
+        if len(memories) == 0:
+            return []
+
+        splice_length = 10
+
+        added_timestamps = []
+        mems_to_filter = []  # Memories without duplicates
+        filtered_mems = []
+
+        # Get the situation
+        queries_txt = "- " + "\n- ".join(queries)
+        context = [make_user_message(situation_prompt.format(queries=queries_txt))]
+        situation = ask_llm(context, model=self.model, max_overall_tokens=self.max_prompt_size, cost_callback=cost_cb, temperature=self.temperature)
+        colour_print("MAGENTA", f"Filtering situation: {situation}")
+
+        # Remove memories that map to the same interaction
+        for m in memories:
+            timestamp = m.timestamp
+            if timestamp not in added_timestamps:
+                added_timestamps.append(timestamp)
+                mems_to_filter.append(m)
+
+        num_splices = ceil(len(mems_to_filter) / splice_length)
+        # Iterate through the mems_to_filter list and create the passage
+        call_count = 0
+        for splice in range(num_splices):
+            start_idx = splice * splice_length
+            end_idx = (splice + 1) * splice_length
+
+            memories_passages = []
+            memory_counter = 0
+
+            for m in mems_to_filter[start_idx:end_idx]:
+                timestamp = m.timestamp
+                um, am = self.session.interaction_from_timestamp(timestamp)
+                memories_passages.append(f"{memory_counter}). (User): {um.content}\n(You): {am.content}\nKeywords: {m.metadata['keywords']}")
+                memory_counter += 1
+
+            queries_txt = "- " + "\n- ".join(queries)
+            passages = "\n----\n".join(memories_passages)
+            context = [make_user_message(prompt.format(passages=passages, situation=situation))]
+            # context = [make_system_message(prompt.format(passages=passages, queries=queries_txt, keywords=keywords))]
+
+            while True:
+                try:
+                    print("Attempting filter")
+                    with open(f"data/llm_calls/{self.session.session_id}/filter-{self.llm_index}-{call_count}.txt", "w") as f:
```

Comment:
We should have this more standardized. I think that using the code that's currently in the baseline, plus session ID and some label should be enough. We could also use a timestamp instead of an index for the file name, to make the code simpler and also make it easier to find the right file.

Reply:
I see. I like the `reply_id/llm_call_id/label` structure. I also don't like the session id very much, but I reckon it's a good safety measure if you want to keep a previous run's logs.

Reply:
I have ditched the `session_id` and have replaced it with a timestamp